### PR TITLE
Add Include/Exclude logic for Attributes Processor

### PR DESCRIPTION
**Pull request by ccaraman**
- Change the internal format of the attributes configuration to have the include and exclude settings in a better internal format
- `matchSpanToProperties` contains the core logic to match a span and service name to a setting. Since this is the same process for include/exclude it is invoked by `shouldProcessSpan` which handles the return value accordingly.
Note: More than half of this PR is test to catch the various possible corner cases.

### CODE REVIEW

File: processor/attributesprocessor/attributes.go [hunk 1]
```diff
@@ -27,10 +27,16 @@ import (
...
-	// This structure is very similar to the config for attributes processor
-	// with the value in the converted attribute format instead of the
-	// raw format from the configuration.
+	config       attributesConfig
+}
+
+// This structure is very similar to the config for attributes processor
+// with the value in the converted attribute format instead of the
+// raw format from the configuration.
+type attributesConfig struct {
```
Comment (tigrannajaryan) [inline processor/attributesprocessor/attributes.go]: 👍

File: processor/attributesprocessor/attributes.go [hunk 2]
```diff
@@ -125,3 +151,110 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
...
+// At least one of services or attributes must be specified. It is supported
+// to have both specified, but both `services` and `attributes` must evaluate
+// to true for a match to occur.
+func matchSpanToProperties(mp matchingProperties, span *tracepb.Span, serviceName string) bool {
+
+	// Check if the service name is in the list of service names.
+	serviceFound := mp.Services[serviceName]
+	// Only return false only when all of these conditions are true:
+	// - a valid service name is provided. Invalid names are empty string,
+	// `<nil-batch-node>`, `<nil-service-info>` and `<empty-service-info-name>`.
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: It is on the next line but i'll make the comment clearer to connect the two sentences - those are the values from `ServiceNameForNode`

File: processor/attributesprocessor/attributes.go [hunk 3]
```diff
@@ -125,3 +151,110 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
 		attributesMap[action.Key] = value
 	}
 }
+
+// shouldProcessSpan determines if a span should be processed.
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: I like `shouldDropSpan` i'll update the code.

File: processor/attributesprocessor/attributes.go [hunk 4]
```diff
@@ -125,3 +151,110 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
...
+}
+
+// matchProperties
+// There are two sets of properties to match against.
+// The service name is checked first, if specified. The attributes are checked
+// afterwards, if specified.
+// At least one of services or attributes must be specified. It is supported
+// to have both specified, but both `services` and `attributes` must evaluate
+// to true for a match to occur.
+func matchSpanToProperties(mp matchingProperties, span *tracepb.Span, serviceName string) bool {
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: I'm going to leave it be at this moment.

File: processor/attributesprocessor/attributes.go [hunk 5]
```diff
@@ -125,3 +151,110 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
...
+
+	// Check if the service name is in the list of service names.
+	serviceFound := mp.Services[serviceName]
+	// Only return false only when all of these conditions are true:
+	// - a valid service name is provided. Invalid names are empty string,
+	// `<nil-batch-node>`, `<nil-service-info>` and `<empty-service-info-name>`.
+	// The last three values are from ServiceNameForNode method.
+	// - the map isn't empty (when the map is empty, serviceFound is always false.)
+	// - serviceFound is false
+	if serviceName != "" &&
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: Ahh - I have a bug here. I don't need to check what the value is of serviceName because by doing that it means the service name check isn't exact and we have cases that by pass it. I'm going to remove this. 

ex the current code (with the bug) 
Trace
```
ServiceName: ""
```

Matching Properties Include
```
Services: ["svcA", "svcB"]
```
The above trace would match which isn't what we want. 

Fixing this and adding the missing test case for it.

File: processor/attributesprocessor/attributes.go [hunk 6]
```diff
@@ -125,3 +151,98 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
...
+			return false
+		}
+	}
+	// If there are no attributes to match against, the span matches.
+	if len(mp.Attributes) == 0 {
+		return true
+	}
+
+	// At this point, it is expected of the span to have attributes because of
+	// len (mp.Attributes) !=0. For spans with no attributes, it does not match.
```
Comment (pjanotti) [inline processor/attributesprocessor/attributes.go]:
```suggestion
	// len (mp.Attributes) != 0. For spans with no attributes, it does not match.
```

File: processor/attributesprocessor/attributes.go [hunk 7]
```diff
@@ -125,3 +151,98 @@ func setAttribute(action attributeAction, attributesMap map[string]*tracepb.Attr
...
+// Include properties are checked before exclude settings are checked.
+func (a *attributesProcessor) shouldDropSpan(span *tracepb.Span, serviceName string) bool {
+	// By default all spans are processed when no include and exclude properties are set.
+	if a.config.include == nil && a.config.exclude == nil {
+		return false
+	}
+
+	if a.config.include != nil {
+		// A false returned in this case means the span should not be processed.
+		if include := matchSpanToProperties(*a.config.include, span, serviceName); !include {
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: Can you elaborate more on what you mean here? maybe with an example?

File: processor/attributesprocessor/attributes.go [hunk 8]
```diff
@@ -44,36 +50,56 @@ type attributeAction struct {
...
 
 func (a *attributesProcessor) ConsumeTraceData(ctx context.Context, td consumerdata.TraceData) error {
+	serviceName := processor.ServiceNameForNode(td.Node)
 	for _, span := range td.Spans {
 		if span == nil {
 			// Do not create empty spans just to add attributes
 			continue
 		}
+
+		if a.shouldDropSpan(span, serviceName) {
```
[the remark this answers is not in the document]
Comment (ccaraman) [reply processor/attributesprocessor/attributes.go]: Oh yeah that is nicer - updating

File: processor/attributesprocessor/attributes.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tigrannajaryan) [reply processor/attributesprocessor/attributes.go]: Sounds good :-)

File: processor/attributesprocessor/attributes.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (pjanotti) [reply processor/attributesprocessor/attributes.go]: sorry, this is fine bad reading on my part: if not included skip, if included and not excluded process it.
